### ISSUE TITLE
Regression: termination checker of Agda-2.5.3 seemingly loops where Agda-2.5.2 passes

### ISSUE THREAD

**Issue by mr-ohman** [agda/agda#2495]
The module Definition.Conversion.Decidable in the codebase below passes for Agda-2.5.2, but loops for Agda-2.6.0-c0386ee until the heap is full:
https://bitbucket.org/mr-ohman/dec-of-conv/commits/4634b1897f6bafab8ec7a293c373fe2c85321ecc

This is almost the same issue as #2458 (same codebase), the only difference is that all modules now use `--without-K` and some changes have been made for the implementation to work without K.

Note that both versions pass if `--without-K` is not used, as below:
https://bitbucket.org/mr-ohman/dec-of-conv/commits/b5284aba50766f8811dedaadf885b3c1a2d1e19e

**Comment by jespercockx** [comment]
Is there any chance you could give a minimal example, or at least a pointer to the file where the problem occurs? That would be very helpful in trying to find out what the problem is.

**Comment by mr-ohman** [comment]
The file where the problem occurs in is Definition/Conversion/Decidable.agda. It is actually enough to remove `{-# OPTIONS --without-K #-}` from that file for type checking to pass in Agda-2.6.0-c0386ee. I will also look into creating a minimal example.

**Comment by jespercockx** [comment]
I can confirm the problem on my machine. The file (and presumably the whole library) checks fine with `--no-termination-check` enabled at the top, so presumably there is some bug in the `--without-K`-specific parts of the termination checker (see comments on #1023).

**Comment by mr-ohman** [comment]
I haven't been able to to create a minimal example of the issue, since the termination checking of the proof passes if some of the larger lemmas are postulated. I was however able to minimize the file where the issue occurs:
https://bitbucket.org/mr-ohman/dec-of-conv/commits/ce0ccc96803913b31cffbd255624466c069b4a66

The minimized variant of Decidable.agda is Definition/Conversion/DecidableIssue2495.agda.

I preformed some additional tests with Agda-2.6.0-b195ba6, a version directly before #2458 was fixed, and found that it loops under the same conditions as this issue with K. For later versions where that issue is fixed the termination checker passes with K.

I suspect that this is the same issue as #2458 but for the `--without-K` variant of the termination checker.

**Comment by jespercockx** [comment]
Thanks for the minimized example, that already helps a lot. The problem is indeed the same as in #2458: the termination checker loops when normalizing call arguments. What's strange is that this is exactly the same code that gets called if `--without-K` is not enabled, so it should also have been fixed by the fix to #2458. @andreasabel could you take a look to see what's different here?

**Comment by andreasabel** [comment]
The mystery is explained by these facts:
1. --without-K prevents with-inlining
2. with-inlining prevents normalization of call arguments

There must be another reason why normalization loops, other than with-inlining.

**Comment by andreasabel** [comment]
Agda loops when normalizing
```
proj₁ (injectivity
                (trans
                 (lemma3 (proj₁ (proj₂ (ne~↓ k~l)))
                  (proj₁ (proj₂ (syntacticEqTerm (soundness~↓ x))))
                  (proj₁ (proj₂ (syntacticEqTerm (soundness~↓ k~l)))))
                 (sym
                  (lemma3 (proj₂ (proj₂ (ne~↓ k~l)))
                   (proj₁ (proj₂ (syntacticEqTerm (soundness~↓ x₂))))
                   (proj₂ (proj₂ (syntacticEqTerm (soundness~↓ k~l))))))))
```
Some of these are postulates: `lemma3`, `ne~↓`, the non-postulates are
```agda
-- If we uncomment any of the following postulates (and comment the
-- corresponding import), the termination checker passes.
-- Otherwise, the termination checker loops for Agda-2.6.0-b195ba6
-- with K (Agda version directly before issue #2458 was fixed)
-- and Agda-2.6.0-c0386ee without K.

-- postulate
--   soundness~↓ : ∀ {k l A Γ} → Γ ⊢ k ~ l ↓ A → Γ ⊢ k ≡ l ∷ A
--   syntacticEqTerm : ∀ {t u A Γ} → Γ ⊢ t ≡ u ∷ A
--                   → Γ ⊢ A × (Γ ⊢ t ∷ A × Γ ⊢ u ∷ A)
--   injectivity : ∀ {Γ F G H E} → Γ ⊢ Π F ▹ G ≡ Π H ▹ E
--               → Γ ⊢ F ≡ H × Γ ∙ F ⊢ G ≡ E

open import Definition.Conversion.Soundness
open import Definition.Typed.Consequences.Syntactic
open import Definition.Typed.Consequences.Injectivity
```

**Comment by andreasabel** [comment]
I shrank this further to the following, in particular, removing the `with` clauses.
```agda
{-# OPTIONS --without-K #-}

module Definition.Conversion.DecidableIssue2495 where

open import Definition.Untyped
open import Definition.Typed
open import Definition.Conversion

open import Tools.Product
open import Tools.Nullary

-- If we uncomment any of the following postulates (and comment the
-- corresponding import), the termination checker passes.
-- Otherwise, the termination checker loops for Agda-2.6.0-b195ba6
-- with K (Agda version directly before issue #2458 was fixed)
-- and Agda-2.6.0-c0386ee without K.

-- postulate
--   soundness~↓ : ∀ {k l A Γ} → Γ ⊢ k ~ l ↓ A → Γ ⊢ k ≡ l ∷ A
--   syntacticEqTerm : ∀ {t u A Γ} → Γ ⊢ t ≡ u ∷ A
--                   → Γ ⊢ A × (Γ ⊢ t ∷ A × Γ ⊢ u ∷ A)
--   injectivity : ∀ {Γ F G H E} → Γ ⊢ Π F ▹ G ≡ Π H ▹ E
--               → Γ ⊢ F ≡ H × Γ ∙ F ⊢ G ≡ E

open import Definition.Conversion.Soundness
open import Definition.Typed.Consequences.Syntactic
open import Definition.Typed.Consequences.Injectivity


postulate
  ANY    : ∀ {a} {A : Set a} → A
  lemma4 : ∀ {t Γ F G} F' G' → Γ ⊢ t ∷ (Π F ▹ G) → Γ ⊢ (Π F ▹ G) ≡ (Π F' ▹ G')
  lemma5 : ∀ {k l Γ} → Dec (Σ Term (_⊢_~_↑_ Γ k l)) → Dec (Σ Term (_⊢_~_↓_ Γ k l))

mutual

  -- needs to be mutual, not a postulate
  decAux :  ∀ t u A B Γ k l
                → Γ ⊢ A ≡ B
                → Dec (∃ λ A → Γ ⊢ k ~ l ↓ A)
                → Dec (∃ λ C → Γ ⊢ k ∘ t ~ l ∘ u ↑ C)
  decAux = ANY

  dec~↑ : ∀ {k l R T Γ}
        → Γ ⊢ k ~ k ↑ R
        → Γ ⊢ l ~ l ↑ T
        → Dec (∃ λ A → Γ ⊢ k ~ l ↑ A)
  dec~↑ {Γ = Γ} (app {k = k} {t = t} {F = F} {G = G} x _)
                (app {k = k'} {t = t'} {F = F'} {G = G'} x₂ _) =
    let ⊢l₁ = proj₁ (proj₂ (syntacticEqTerm (soundness~↓ x)))
        F≡F' = proj₁ (injectivity {Γ} {F} {G} {F'} {G'} (lemma4 F' G' ⊢l₁))
    in  {! decAux t t' F F' Γ k k' F≡F' (decDown x x₂) !}  -- F≡F' needed
  dec~↑ _ = ANY

  decDown : ∀ {k l R T Γ}
        → Γ ⊢ k ~ k ↓ R
        → Γ ⊢ l ~ l ↓ T
        → Dec (∃ λ A → Γ ⊢ k ~ l ↓ A)
  decDown ([~] A D whnfB k~l) ([~] A₁ D₁ whnfB₁ k~l₁) = lemma5 (dec~↑ k~l k~l₁)   -- NEEDED
```
If you give the expression, the termination checker loops.

**Comment by andreasabel** [comment]
After more shrinking I now think that Agda does not loop, but just does not finish normalizing.  
I am not sure how to address this, maybe remove new feature #2403, or only reduce non-recursive functions.

**Comment by WolframKahl** [comment]
I also have a module (on top of a large development) that typechecked with 2.5.1 and 7c5f712aac3afefeb0a05329592b263437a40b98 in at most a few minutes, but fails to typecheck in days with 2.5.2 and current stable. (However, it typechecks quickly with @Saizan 's patch from #1625 .

**Comment by nad** [comment]
@WolframKahl, perhaps you could use `agda-bisect` to try to track down some commit that made Agda considerably slower for your example.

Perhaps you could use a script of the following form:
```
#!/bin/sh

AGDA=$1

# Make sure that all the dependencies have been checked.
$AGDA <suitable options> <some module that imports the same modules as your problematic module>

timeout <suitable duration, a bit more than "a few minutes"> $AGDA <suitable options> <problematic module>
```

A suggested command-line (untested):
```
$ agda-bisect --good v2.5.1 --bad v2.5.2 --log log --script ./script
```
I suggest that you use `--dry-run` to make sure that the script does what it should for Agda 2.5.1 and Agda 2.5.2.

**Comment by WolframKahl** [comment]
I'll try the bisection, perhaps on the week-end. Thank you , @nad !

Slight complication: A language change somewhere in the middle of that range...

**Comment by WolframKahl** [comment]
Observation: There is a tag `2.5.1`, but no tag `2.5.2`.
````
 $ git show 2.5.2 --
fatal: bad revision '2.5.2'
````

**Comment by asr** [comment]
For 2.5.2 the tag name is `v2.5.2`.

**Comment by andreasabel** [comment]
Fixed this by not reducing recursive functions during normalization of arguments in the termination checker.

**Comment by nad** [comment]
A rule of thumb is to avoid normalisation as much as possible. What do you mean by "expand"?

I don't think that the problem brought up in #2403 needs to be fixed, and "expand" sounds unprincipled to me (users might find it hard to understand why the code works in some cases and not in others). Perhaps we could skip the expansion.

**Comment by andreasabel** [comment]
By expand I mean unfold, i.e., apply the reduction rules.

The termination checker now does a bit of (hopefully inexpensive) normalization on the arguments of the recursive call before doing the structural comparison with the function parameters.  

Maybe that could be done in a smarter way.  

Anyway I think the discussion is about #2403, maybe it should be continued there.  I close this issue.

**Comment by andreasabel** [comment]
The fix here might have introduced regressions #4447 and #2901.
